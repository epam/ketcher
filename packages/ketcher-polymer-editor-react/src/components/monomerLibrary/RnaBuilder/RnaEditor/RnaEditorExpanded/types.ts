import { ChangeEvent } from 'react';

export interface IRnaEditorExpandedProps {
  name: string;
  onCancel: () => void;
  onChangeName: (event: ChangeEvent<HTMLInputElement>) => void;
}
