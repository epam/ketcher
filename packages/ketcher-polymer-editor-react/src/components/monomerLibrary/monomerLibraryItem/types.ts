/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { Struct } from 'ketcher-core';
import { MonomerColorScheme } from 'theming/defaultTheme';

export type MonomerItemType = {
  label: string;
  colorScheme?: MonomerColorScheme;
  favorite?: boolean;
  struct: Struct;
  props: {
    MonomerNaturalAnalogCode: string;
    MonomerName: string;
    Name: string;
    // TODO determine whenever these props are optional or not
    BranchMonomer?: string;
    MonomerCaps?: string;
    MonomerCode?: string;
    MonomerType?: string;
  };
};

export interface IMonomerItemProps {
  item: MonomerItemType;
  isSelected?: boolean;
  onClick?: VoidFunction;
  onMouseLeave?: (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
  onMouseMove?: (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
  onStarClick?: VoidFunction;
}
