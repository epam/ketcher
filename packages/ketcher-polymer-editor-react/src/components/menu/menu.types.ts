/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

type MenuItemVariant =
  | 'open'
  | 'select-lasso'
  | 'select-rectangle'
  | 'select-fragment'
  | 'rectangle'
  | 'ellipse'
  | 'help'
  | 'settings'
  | 'undo'
  | 'single-bond'
  | 'erase'
  | 'rotate'
  | 'horizontal-flip'
  | 'vertical-flip'
  | 'bracket'

export type { MenuItemVariant }
