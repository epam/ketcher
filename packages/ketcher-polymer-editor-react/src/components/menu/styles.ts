/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import styled from '@emotion/styled';

export const MenuLayout = styled.div`
  background-color: ${({ theme }) => theme.ketcher.color.background.primary};
`;

export const Divider = styled.span`
  display: block;
  height: 8px;
  width: 32px;
  border-top: 1px solid;
  border-color: ${({ theme }) => theme.ketcher.color.divider};
`;

export const StyledGroup = styled.div`
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  align-items: center;
  background-color: ${({ theme }) => theme.ketcher.color.background.primary};
  border-radius: 2px;
  width: 32px;
  margin-bottom: 8px;

  > * {
    margin-bottom: 8px;
  }

  > :last-child {
    margin-bottom: 0;
  }
`;
