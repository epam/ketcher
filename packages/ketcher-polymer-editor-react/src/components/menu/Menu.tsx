/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import React from 'react';
import { MenuItem } from './menuItem';
import { SubMenu } from './subMenu';
import { IMenuContext, MenuContext } from '../../contexts';
import { Divider, MenuLayout, StyledGroup } from './styles';
import { GroupProps, MenuProps } from './types';

const Group = ({
  children,
  divider = false,
}: React.PropsWithChildren<GroupProps>) => {
  const subComponents = React.Children.map(
    children as JSX.Element[],
    (child) => {
      return child.type === MenuItem || child.type === SubMenu ? child : null;
    },
  );

  return (
    <>
      <StyledGroup>{subComponents.map((component) => component)}</StyledGroup>
      {divider && <Divider />}
    </>
  );
};

const Menu = ({
  children,
  onItemClick,
  activeMenuItem,
}: React.PropsWithChildren<MenuProps>) => {
  const context = React.useMemo<IMenuContext>(
    () => ({
      isActive: (itemKey) => activeMenuItem === itemKey,
      activate: (itemKey) => {
        onItemClick(itemKey);
      },
    }),
    [activeMenuItem, onItemClick],
  );

  const subComponents = React.Children.map(
    children as JSX.Element[],
    (child) => {
      return child.type === Group ? child : null;
    },
  );

  return (
    <MenuContext.Provider value={context}>
      <MenuLayout>{subComponents.map((component) => component)}</MenuLayout>
    </MenuContext.Provider>
  );
};

Menu.Group = Group;
Menu.Item = MenuItem;
Menu.Submenu = SubMenu;

export { Menu, MenuContext };
