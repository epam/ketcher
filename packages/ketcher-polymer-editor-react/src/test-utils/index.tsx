import { createTheme } from '@mui/material/styles'
import {
  render as rtlRender,
  RenderOptions,
  RenderResult
} from '@testing-library/react'
import React, { ReactElement } from 'react'
import { Provider } from 'react-redux'
import { ThemeProvider } from '@emotion/react'
import { merge } from 'lodash'

import { configureAppStore, RootState } from 'state'
import { defaultTheme } from 'theming/defaultTheme'

const muiTheme = createTheme()

interface CustomRenderOptions {
  preloadedState?: RootState
  renderOptions?: Omit<RenderOptions, 'wrapper'>
}

function render(
  ui: ReactElement,
  { preloadedState = {}, ...renderOptions }: CustomRenderOptions = {}
): RenderResult {
  const Wrapper: React.FC = ({ children }) => {
    const store = configureAppStore(preloadedState)
    return (
      <ThemeProvider theme={merge(muiTheme, { ketcher: defaultTheme })}>
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    )
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// eslint-disable-next-line import/export
export * from '@testing-library/react'

// eslint-disable-next-line import/export
export { render }
