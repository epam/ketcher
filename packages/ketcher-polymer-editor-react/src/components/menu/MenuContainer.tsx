/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { useAppDispatch, useAppSelector } from 'hooks'
import { selectEditorActiveTool, selectTool } from 'state/common'
import { Menu } from 'components/menu/Menu'

const MenuContainer = () => {
  const dispatch = useAppDispatch()
  const activeTool = useAppSelector(selectEditorActiveTool)

  const menuProps = {
    activeItem: activeTool,
    onClick: (tool) => dispatch(selectTool(tool))
  }
  return (
    <Menu>
      <Menu.Group>
        <Menu.Item name="open" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="undo" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="erase" {...menuProps} />
        <Menu.Submenu vertical {...menuProps}>
          <Menu.Item name="select-lasso" {...menuProps} />
          <Menu.Item name="select-rectangle" {...menuProps} />
          <Menu.Item name="select-fragment" {...menuProps} />
        </Menu.Submenu>
        <Menu.Submenu {...menuProps}>
          <Menu.Item name="rectangle" {...menuProps} />
          <Menu.Item name="ellipse" {...menuProps} />
        </Menu.Submenu>
        <Menu.Submenu {...menuProps}>
          <Menu.Item name="rotate" {...menuProps} />
          <Menu.Item name="horizontal-flip" {...menuProps} />
          <Menu.Item name="vertical-flip" {...menuProps} />
        </Menu.Submenu>
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="single-bond" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="bracket" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="settings" {...menuProps} />
        <Menu.Item name="help" {...menuProps} />
      </Menu.Group>
    </Menu>
  )
}
export { MenuContainer }
