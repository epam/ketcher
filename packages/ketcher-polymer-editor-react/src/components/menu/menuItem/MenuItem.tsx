/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { Icon } from 'components/shared/ui/icon'
import { MenuItemVariant } from 'components/menu/menu.types'
import { useContext } from 'react'
import { MenuContext } from 'components/menu'
import { MenuItem as MuiMenuItem } from '@mui/material'
import { css, useTheme } from '@emotion/react'

const CustomMenuItem = ({ isActive, onClick, children }) => {
  const theme = useTheme()
  const styles = css`
    display: flex;
    align-items: center;
    width: 32px;
    height: 32px;
    padding: 0;
    justify-content: center;
    border-radius: 2px;
    background-color: ${isActive
      ? theme.color.icon.activeMenu
      : theme.color.background.primary};

    &:hover {
      background: ${isActive
        ? theme.color.icon.activeMenu
        : theme.color.background.primary};
    }

    & > svg path {
      fill: ${isActive
        ? theme.color.icon.clicked
        : theme.color.icon.activeMenu};
    }
  `
  return (
    <MuiMenuItem css={styles} onClick={onClick}>
      {children}
    </MuiMenuItem>
  )
}

type MenuItemProp = {
  itemKey: MenuItemVariant
}

const MenuItem = ({ itemKey }: MenuItemProp) => {
  const { isActiveItem, itemClickHandler } = useContext(MenuContext)

  return (
    <CustomMenuItem
      isActive={isActiveItem(itemKey)}
      onClick={() => itemClickHandler(itemKey)}
    >
      <Icon name={itemKey} />
    </CustomMenuItem>
  )
}

export { MenuItem }
