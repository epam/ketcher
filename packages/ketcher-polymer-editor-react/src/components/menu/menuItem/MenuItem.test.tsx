/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { render, screen } from 'test-utils'
import { MenuItem } from 'components/menu/menuItem/MenuItem'
import { MenuItemVariant } from 'components/menu/menu.types'
import userEvent from '@testing-library/user-event'

const mockClickHandler = jest.fn()
const MOCK_NAME: MenuItemVariant = 'select-lasso'

const mockProps = {
  name: MOCK_NAME,
  onClick: mockClickHandler,
  activeItem: MOCK_NAME
}

describe('Test MenuItem component', () => {
  it('should be rendered without crashing', () => {
    const { asFragment } = render(<MenuItem {...mockProps} />)
    expect(asFragment).toMatchSnapshot()
  })
  it('should render menu icon element when props are provided', () => {
    render(<MenuItem {...mockProps} />)
    expect(screen.getByRole('button')).toBeInTheDocument()
  })
  it('should call provided callback when menu icon is clicked', () => {
    render(<MenuItem {...mockProps} />)
    const button = screen.getByRole('button')
    userEvent.click(button)
    expect(mockClickHandler).toHaveBeenCalledTimes(1)
  })
})
