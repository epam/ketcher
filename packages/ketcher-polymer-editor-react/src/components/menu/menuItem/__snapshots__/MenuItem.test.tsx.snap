// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MenuItem component should be rendered without crashing 1`] = `[Function]`;
